pkg/ambex/main.go

```diff
@@ -802,114 +801,107 @@ func Main(
-	go func() {
-		defer close(envoyUpdaterDone)
-		if err := Updater(ctx, updates, getUsage); err != nil {
-			panic(err) // TODO: Find a better way of reporting errors from goroutines.
+	grp.Go("updater", func(ctx context.Context) error {
+		return Updater(ctx, updates, getUsage)
+	})
+	grp.Go("main-loop", func(ctx context.Context) error {
+		generation := 0
+		var fastpathSnapshot *FastpathSnapshot
+		edsEndpoints := map[string]*v2.ClusterLoadAssignment{}
+		edsEndpointsV3 := map[string]*v3endpointconfig.ClusterLoadAssignment{}
+
+		// We always start by updating with a totally empty snapshot.
+		//
+		// XXX This seems questionable: why do we do this? Envoy isn't currently started until
+		// we have a real configuration...
+		err = update(
+			ctx,
+			args.snapdirPath,
+			args.numsnaps,
+			config,
+			configv3,
+			&generation,
+			args.dirs,
+			edsEndpoints,
+			edsEndpointsV3,
+			fastpathSnapshot,
+			updates,
+		)
+		if err != nil {
+			return err
 		}
-	}()
-
-	generation := 0
-	var fastpathSnapshot *FastpathSnapshot
-	edsEndpoints := map[string]*v2.ClusterLoadAssignment{}
-	edsEndpointsV3 := map[string]*v3endpointconfig.ClusterLoadAssignment{}
-
-	// We always start by updating with a totally empty snapshot.
-	//
-	// XXX This seems questionable: why do we do this? Envoy isn't currently started until
-	// we have a real configuration...
-	err = update(
-		ctx,
-		args.snapdirPath,
-		args.numsnaps,
-		config,
-		configv3,
-		&generation,
-		args.dirs,
-		edsEndpoints,
-		edsEndpointsV3,
-		fastpathSnapshot,
-		updates,
-	)
-	if err != nil {
-		return err
-	}
 
-	// This is the main loop where the magic happens. The fact that it uses a label
-	// depresses me, though.
-OUTER:
-	for {
-
-		select {
-		case _ = <-sigCh:
-			err := update(
-				ctx,
-				args.snapdirPath,
-				args.numsnaps,
-				config,
-				configv3,
-				&generation,
-				args.dirs,
-				edsEndpoints,
-				edsEndpointsV3,
-				fastpathSnapshot,
-				updates,
-			)
-			if err != nil {
-				return err
-			}
-		case fpSnap := <-fastpathCh:
-			// Fastpath update. Grab new endpoints and update.
-			if fpSnap.Endpoints != nil {
-				edsEndpoints = fpSnap.Endpoints.ToMap_v2()
-				edsEndpointsV3 = fpSnap.Endpoints.ToMap_v3()
-			}
-			fastpathSnapshot = fpSnap
-			err := update(
-				ctx,
-				args.snapdirPath,
-				args.numsnaps,
-				config,
-				configv3,
-				&generation,
-				args.dirs,
-				edsEndpoints,
-				edsEndpointsV3,
-				fastpathSnapshot,
-				updates,
-			)
-			if err != nil {
-				return err
-			}
-		case <-watcher.Events:
-			// Non-fastpath update. Just update.
-			err := update(
-				ctx,
-				args.snapdirPath,
-				args.numsnaps,
-				config,
-				configv3,
-				&generation,
-				args.dirs,
-				edsEndpoints,
-				edsEndpointsV3,
-				fastpathSnapshot,
-				updates,
-			)
-			if err != nil {
-				return err
+		// This is the main loop where the magic happens. The fact that it uses a label
+		// depresses me, though.
```

Comment:
Oh good, it doesn't use a label any more. Maybe change the comment?

Reply:
I'll address that in a follow-up PR.